pyrokinetics/gk_code/GKOutputReaderCGYRO.py

```diff
@@ -0,0 +1,406 @@
+import xarray as xr
+import logging
+from typing import Tuple, Dict, Any, Optional
+from pathlib import Path
+
+from .GKOutputReader import GKOutputReader
+from .GKInputCGYRO import GKInputCGYRO
+from ..constants import pi
+from ..typing import PathLike
+
+
+class CGYROFile:
+    def __init__(self, path: PathLike, required: bool):
+        self.path = Path(path)
+        self.required = required
+        self.fmt = self.path.name.split(".")[0]
+
+
+class GKOutputReaderCGYRO(GKOutputReader):
+
+    fields = ["phi", "apar", "bpar"]
+
+    @staticmethod
+    def _required_files(dirname: PathLike):
+        dirname = Path(dirname)
+        return {
+            "input": CGYROFile(dirname / "input.cgyro", required=True),
+            "time": CGYROFile(dirname / "out.cgyro.time", required=True),
+            "grids": CGYROFile(dirname / "out.cgyro.grids", required=True),
+            "equilibrium": CGYROFile(dirname / "out.cgyro.equilibrium", required=True),
+        }
+
+    def verify(self, dirname: PathLike):
+        dirname = Path(dirname)
+        for f in self._required_files(dirname).values():
+            if not f.path.exists():
+                raise RuntimeError
+
+    @classmethod
+    def _get_raw_data(
+        cls, dirname: PathLike
+    ) -> Tuple[Dict[str, Any], GKInputCGYRO, str]:
+        dirname = Path(dirname)
+        if not dirname.exists():
+            raise RuntimeError(
+                f"GKOutputReaderCGYRO: Provided path {dirname} does not exist. "
+                "Please supply the name of a directory containing CGYRO output files."
+            )
+        if not dirname.is_dir():
+            raise RuntimeError(
+                f"GKOutputReaderCGYRO: Provided path {dirname} is not a directory. "
+                "Please supply the name of a directory containing CGYRO output files."
+            )
+
+        # The following list of CGYRO files may exist
+        expected_files = {
+            **cls._required_files(dirname),
+            "flux": CGYROFile(dirname / "bin.cgyro.ky_flux", required=False),
+            "eigenvalues_bin": CGYROFile(dirname / "bin.cgyro.freq", required=False),
+            "eigenvalues_out": CGYROFile(dirname / "out.cgyro.freq", required=False),
+            **{
+                f"field_{f}": CGYROFile(dirname / f"bin.cgyro.kxky_{f}", required=False)
+                for f in cls.fields
+            },
+            **{
+                f"eigenfunctions_{f}": CGYROFile(
+                    dirname / f"bin.cgyro.{f}b", required=False
+                )
+                for f in cls.fields
+            },
+        }
+        # Read in files
+        raw_data = {}
+        for key, cgyro_file in expected_files.items():
+            if not cgyro_file.path.exists():
+                if not cgyro_file.required:
+                    continue
+                raise RuntimeError(
+                    f"GKOutputReaderCGYRO: The file {cgyro_file.path.name} is needed"
+                )
```

Comment:
This is perhaps a bit clearer:

```suggestion
                if cgyro_file.required:
                    raise RuntimeError(
                        f"GKOutputReaderCGYRO: The file {cgyro_file.path.name} is needed"
                    )
```